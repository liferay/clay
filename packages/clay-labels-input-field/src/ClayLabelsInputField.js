import 'clay-button';
import 'clay-label';
import {Config} from 'metal-state';
import {EventHandler} from 'metal-events';
import {match} from './utils';
import ClayComponent from 'clay-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';

import templates from './ClayLabelsInputField.soy.js';

/**
 * Metal ClayLabelsInputField component.
 * @extends ClayComponent
 */
class ClayLabelsInputField extends ClayComponent {
	/**
	 * Get the value typed in the input or through the filtered items.
	 * @param {!Event} event
	 * @protected
	 * @return {string}
	 */
	_getLabelFromEvent(event) {
		const {tagName} = event.target;

		if (tagName === 'INPUT') {
			return event.target.value.toLowerCase().replace(',', '');
		} else if (tagName === 'A') {
			const index = event.target.getAttribute('data-dropdown-item-index');
			const element = this.filteredItems[Number(index)];

			return element.originalString.toLowerCase();
		}
	}

	/**
	 * Continues the propagation of the Button clicked event.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleButtonClicked(event) {
		return !this.emit({
			name: 'buttonClicked',
			originalEvent: event,
		});
	}

	/**
	 * Handle the click on the close label button.
	 * @param {!Object} data
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleCloseButtonClick(data) {
		return this._handleLabelRemoved(data.target.element);
	}

	/**
	 * Handle the click on the dropdown item and the propagation of the labelAdded event.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleDropdownItemClick(event) {
		return this._handleLabelAdded(event);
	}

	/**
	 * Continues the propagation of the labelAdded event.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleLabelAdded(event) {
		const label = this._getLabelFromEvent(event);

		if (
			label.trim() &&
			!this.selectedLabels
				.find((labelSelected) => labelSelected.label === label)
		) {
			return !this.emit({
				data: {
					value: label,
				},
				name: 'labelAdded',
				originalEvent: event,
			});
		} else {
			this.refs.input.value = label;
		}
	}

	/**
	 * Handle the focus on the selected labels and triggers the focused label event.
	 * @param {!Event} event
	 * @param {?Boolean} direction
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleLabelFocus(event, direction) {
		if (this.selectedLabels.length) {
			const {formGroupInput} = this.refs;
			const items = formGroupInput.querySelectorAll('span[id="item-tag"]');

			if (this._labelFocused) {
				const index = this._labelFocused.getAttribute('data-tag');
				const condition = direction
					? Number(index) - 1
					: Number(index) + 1;

				if (condition > (items.length - 1)) {
					this.refs.input.focus();
					this._removeFocusedLabel();
					return false;
				} else if (condition >= 0) {
					this._labelFocused.classList.remove('label-focus');
					this._labelFocused = items[condition];
					this._labelFocused.classList.add('label-focus');
				} else {
					return false;
				}
			} else {
				this._labelFocused = items[items.length - 1];
				this._labelFocused.classList.toggle('label-focus');
			}

			return !this.emit({
				data: {
					item: this._labelFocused,
				},
				name: 'labelFocused',
				originalEvent: event,
			});
		}
	}

	/**
	 * Handle the focused label removal and propagating the labelRemoved event.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleLabelRemoved(event) {
		const index = event.getAttribute('data-tag');

		this._removeFocusedLabel();

		return !this.emit({
			data: {
				index,
			},
			name: 'labelRemoved',
			originalEvent: event,
		});
	}

	/**
	 * Handles input changes and propagates the queryChange event.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleOnInput(event) {
		const value = event.target.value.toLowerCase();

		this._removeFocusedLabel();

		switch (event.data) {
		case ',':
			return this._handleLabelAdded(event);
		default:
			return !this.emit({
				data: {
					value,
				},
				name: 'queryChange',
				originalEvent: event,
			});
		}
	}

	/**
	 * Handles form interactions and propagates corresponding events.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleOnKeydown(event) {
		switch (event.key) {
		case 'Enter':
			event.preventDefault();
			if (this._labelFocused) {
				return this._handleLabelRemoved(this._labelFocused);
			} else {
				return this._handleLabelAdded(event);
			}
		case 'Backspace':
			if (!this.refs.input.value) {
				if (!this._labelFocused) {
					return this._handleLabelFocus(event);
				} else {
					return this._handleLabelRemoved(this._labelFocused);
				}
			}
			break;
		case 'ArrowLeft':
			if (!this.refs.input.value && this._labelFocused) {
				return this._handleLabelFocus(event, true);
			}
			break;
		case 'ArrowRight':
			if (!this.refs.input.value && this._labelFocused) {
				return this._handleLabelFocus(event, false);
			}
			break;
		case 'ArrowUp':
			this._setFocusItemDropdown(true);
			break;
		case 'ArrowDown':
			this._setFocusItemDropdown(false);
			break;
		}
	}

	/**
	 * Removes the focus from the focused element.
	 * @protected
	 */
	_removeFocusedLabel() {
		if (this._labelFocused) {
			this._labelFocused.classList.remove('label-focus');
			this._labelFocused = null;
		}
	}

	/**
	 * Handle the interactions in the dropdown and add focus on the items.
	 * @protected
	 * @param {!Boolean} direction
	 */
	_setFocusItemDropdown(direction) {
		if (this.filteredItems.length) {
			const elements = this.refs.dropdown.querySelectorAll('a[id="item"]');

			if (direction && this._dropdownItemFocused === 0) {
				this.refs.input.focus();
				this._dropdownItemFocused = null;
			} else {
				this._dropdownItemFocused =
					this._dropdownItemFocused === null
					|| (elements.length - 1) === this._dropdownItemFocused
						? 0
						: direction
							? this._dropdownItemFocused - 1
							: this._dropdownItemFocused + 1;

				elements[this._dropdownItemFocused].focus();
			}
		}
	}

	/**
	 * @inheritDoc
	 */
	attached() {
		this._labelFocused = null;
		this._dropdownItemFocused = null;
	}

	/**
	 * Clears the input value.
	 * @public
	 */
	clearInput() {
		this.refs.input.value = '';
	}

	/**
	 * @inheritDoc
	 */
	created() {
		this._eventHandler = new EventHandler();
	}

	/**
	 * @inheritDoc
	 */
	disposed() {
		this._eventHandler.removeAllListeners();
	}

	/**
	 * Helper method to filter a list based on a string.
	 * @param {!Array<string>} data
	 * @param {!string} query
	 * @public
	 * @return {Array} A list of items containing the corresponding characters
	 */
	filterAutocomplete(data, query) {
		return data
			.reduce((prev, element, index) => {
				let result = match(query, element);

				if (result != null) {
					prev[prev.length] = {
						index,
						matches: result.values,
						originalString: element,
						score: result.score,
					};
				}

				return prev;
			}, [])
			.sort((a, b) => {
				if (a > b) return 1;
				if (a < b) return -1;
			});
	}

	/**
	 * @inheritDoc
	 */
	syncFilteredItems() {
		this._dropdownItemFocused = null;
	}

	/**
	 * @inheritDoc
	 */
	syncSelectedLabels() {
		this.refs.input.focus();
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayLabelsInputField.STATE = {
	/**
	 * Variation name to render different deltemplates.
	 * @default undefined
	 * @instance
	 * @memberof ClayLabelsInputField
	 * @type {?(string|undefined)}
	 */
	contentRenderer: Config.string(),

	/**
	 * CSS classes to be applied to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayLabelsInputField
	 * @type {?(string|undefined)}
	 */
	elementClasses: Config.string(),

	/**
	 * List of filtered items for suggestion or autocomplete.
	 * @default []
	 * @instance
	 * @memberof ClayLabelsInputField
	 * @type {?Array}
	 */
	filteredItems: Config.array(Config.object()).value([]),

	/**
	 * Help text to guide the user in the interaction.
	 * @default undefined
	 * @instance
	 * @memberof ClayLabelsInputField
	 * @type {!string}
	 */
	helpText: Config.string().required(),

	/**
     * Id to be applied to the element.
     * @default undefined
     * @instance
     * @memberof ClayLabelsInputField
     * @type {?(string|undefined)}
     */
	id: Config.string(),

	/**
	 * Label of the input element.
	 * @default undefined
	 * @instance
	 * @memberof ClayLabelsInputField
	 * @type {?(string|undefined)}
	 */
	label: Config.string(),

	/**
	 * List of the selected Labels.
	 * @default []
	 * @instance
	 * @memberof ClayLabelsInputField
	 * @type {?Array<Object>}
	 */
	selectedLabels: Config.array(Config.object()).value([]),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @default undefined
	 * @instance
	 * @memberof ClayLabelsInputField
	 * @type {!string}
	 */
	spritemap: Config.string().required(),
};

defineWebComponent('clay-labels-input-field', ClayLabelsInputField);

Soy.register(ClayLabelsInputField, templates);

export {ClayLabelsInputField};
export default ClayLabelsInputField;
