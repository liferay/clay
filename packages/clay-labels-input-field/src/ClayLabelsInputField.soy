{namespace ClayLabelsInputField}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param spritemap: string}
	{@param? _handleButtonClicked: any}
	{@param? _handleCloseButtonClick: any}
	{@param? _handleDropdownItemClick: any}
	{@param? _handleOnInput: any}
	{@param? _handleOnKeydown: any}
	{@param? _removeFocusedLabel: any}
	{@param? contentRenderer: string}
	{@param? elementClasses: string}
	{@param? filteredItems: list<?>}
	{@param? helpText: string}
	{@param? id: string}
	{@param? label: string}
	{@param? selectedLabels: list<?>}

	{let $attributes kind="attributes"}
		class="from-group
			{if $elementClasses}
				{sp}{$elementClasses}
			{/if}
		"

		{if $id}
			id="{$id}"
		{/if}
	{/let}

	<div {$attributes}>
		{if $label}
			<label>{$label}</label>
		{/if}

		{call .content}
			{param _handleButtonClicked: $_handleButtonClicked /}
			{param _handleCloseButtonClick: $_handleCloseButtonClick /}
			{param _handleDropdownItemClick: $_handleDropdownItemClick /}
			{param _handleOnInput: $_handleOnInput /}
			{param _handleOnKeydown: $_handleOnKeydown /}
			{param _removeFocusedLabel: $_removeFocusedLabel /}
			{param contentRenderer: $contentRenderer /}
			{param filteredItems: $filteredItems /}
			{param helpText: $helpText /}
			{param selectedLabels: $selectedLabels /}
			{param spritemap: $spritemap /}
		{/call}
	</div>
{/template}

{template .content}
	{@param spritemap: string}
	{@param? _handleButtonClicked: any}
	{@param? _handleCloseButtonClick: any}
	{@param? _handleDropdownItemClick: any}
	{@param? _handleOnInput: any}
	{@param? _handleOnKeydown: any}
	{@param? _removeFocusedLabel: any}
	{@param? contentRenderer: string}
	{@param? filteredItems: list<?>}
	{@param? helpText: string}
	{@param? selectedLabels: list<?>}

	<div class="input-group input-group-stacked-sm-down">
		<div class="input-group-item">
			<div data-keydown="{$_handleOnKeydown}" ref="formGroupInput" class="form-control dropdown-full form-control-tag-group">
				{if $selectedLabels}
					{foreach $item in $selectedLabels}
						{delcall ClayInputField.Label variant="$contentRenderer" allowemptydefault="true"}
							{param _handleCloseButtonClick: $_handleCloseButtonClick /}
							{param data: $item /}
							{param index: index($item) /}
							{param spritemap: $spritemap /}
						{/delcall}
					{/foreach}
				{/if}

				<input
					class="form-control-inset"
					onBlur="{$_removeFocusedLabel}"
					onFocus="{$_removeFocusedLabel}"
					onInput="{$_handleOnInput}"
					onKeydown="{$_handleOnKeydown}"
					ref="input"
					type="text"
				/>

				{let $isVisible: $filteredItems and length($filteredItems) > 0 ? true : false /}

				{let $classes kind="text"}
					dropdown-menu
					{if $isVisible}
						{sp}show
					{/if}
				{/let}

				<ul onkeydown="{$_handleOnKeydown}" class="{$classes}" ref="dropdown">
					{if $filteredItems}
						{foreach $item in $filteredItems}
							{delcall ClayInputField.ItemDropdown variant="$contentRenderer" allowemptydefault="true"}
								{param _handleDropdownItemClick: $_handleDropdownItemClick /}
								{param data: $item /}
								{param index: index($item) /}
							{/delcall}
						{/foreach}
					{/if}
				</ul>
			</div>

			{if $helpText and not $isVisible}
				<div class="form-feedback-group">
					<div class="form-text">
						{$helpText}
					</div>
				</div>
			{/if}
		</div>
		<div class="input-group-item input-group-item-shrink">
			{call ClayButton.render}
				{param events: ['click': $_handleButtonClicked] /}
				{param label kind="text"}
					{msg desc="Select items"}
						select
					{/msg}
				{/param}

				{param style: 'secondary' /}
			{/call}
		</div>
	</div>
{/template}

{deltemplate ClayInputField.Label}
	{@param data: ?}
	{@param index: int}
	{@param spritemap: string}
	{@param? _handleCloseButtonClick: any}

	{call ClayLabel.render}
		{param closeable: true /}
		{param data: ['tag': $index] /}
		{param events: ['close': $_handleCloseButtonClick] /}
		{param id: 'item-tag' /}
		{param label: $data.label /}
		{param spritemap: $spritemap /}
	{/call}
{/deltemplate}

{deltemplate ClayInputField.ItemDropdown}
	{@param data: ?}
	{@param index: number}
	{@param? _handleDropdownItemClick: any}

	<li>
		<a data-onclick="{$_handleDropdownItemClick}" data-dropdown-item-index="{$index}" id="item" class="dropdown-item" href="javascript:;">
			{if $data.matches and length($data.matches) > 0}
				{foreach $char in $data.matches}
					{if $char.match}
						<strong>{$char.value}</strong>
					{else}
						{$char.value}
					{/if}
				{/foreach}
			{/if}
		</a>
	</li>
{/deltemplate}
