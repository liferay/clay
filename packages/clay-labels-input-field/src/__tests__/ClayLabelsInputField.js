import ClayLabelsInputField from '../ClayLabelsInputField';

let component;
let spritemap = 'icons.svg';
let helpText = 'You can use a comma to enter tags';

describe('ClayLabelsInputField', function() {
	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render the default markup', () => {
		component = new ClayLabelsInputField({
			helpText,
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayLabelsInputField with classes', () => {
		component = new ClayLabelsInputField({
			elementClasses: 'my-custom-class',
			helpText,
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayLabelsInputField with id', () => {
		component = new ClayLabelsInputField({
			helpText,
			id: 'myId',
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});
});
