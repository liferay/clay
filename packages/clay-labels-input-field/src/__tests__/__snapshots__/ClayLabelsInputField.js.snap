// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayLabelsInputField should render a ClayLabelsInputField with classes 1`] = `
<div class="from-group my-custom-class">
  <div class="input-group input-group-stacked-sm-down">
    <div class="input-group-item">
      <div ref="formGroupInput" class="form-control dropdown-full form-control-tag-group">
        <input class="form-control-inset" ref="input" type="text">
        <ul class="dropdown-menu" ref="dropdown"></ul>
      </div>
      <div class="form-feedback-group">
        <div class="form-text">You can use a comma to enter tags</div>
      </div>
    </div>
    <div class="input-group-item input-group-item-shrink">
      <button class="btn btn-secondary" type="button">select</button>
    </div>
  </div>
</div>
`;

exports[`ClayLabelsInputField should render a ClayLabelsInputField with id 1`] = `
<div class="from-group" id="myId">
  <div class="input-group input-group-stacked-sm-down">
    <div class="input-group-item">
      <div ref="formGroupInput" class="form-control dropdown-full form-control-tag-group">
        <input class="form-control-inset" ref="input" type="text">
        <ul class="dropdown-menu" ref="dropdown"></ul>
      </div>
      <div class="form-feedback-group">
        <div class="form-text">You can use a comma to enter tags</div>
      </div>
    </div>
    <div class="input-group-item input-group-item-shrink">
      <button class="btn btn-secondary" type="button">select</button>
    </div>
  </div>
</div>
`;

exports[`ClayLabelsInputField should render the default markup 1`] = `
<div class="from-group">
  <div class="input-group input-group-stacked-sm-down">
    <div class="input-group-item">
      <div ref="formGroupInput" class="form-control dropdown-full form-control-tag-group">
        <input class="form-control-inset" ref="input" type="text">
        <ul class="dropdown-menu" ref="dropdown"></ul>
      </div>
      <div class="form-feedback-group">
        <div class="form-text">You can use a comma to enter tags</div>
      </div>
    </div>
    <div class="input-group-item input-group-item-shrink">
      <button class="btn btn-secondary" type="button">select</button>
    </div>
  </div>
</div>
`;
